### PR TITLE
fix: returning correct http status code in case of failures

### PR DESCRIPTION
**Pull request by prashbnair**
Fixes  https://github.com/prometheus-community/prom-label-proxy/issues/51

Signed-off-by: Prashant Balachandran <pnair@redhat.com>

### CODE REVIEW

File: injectproxy/routes.go [hunk 1]
```diff
@@ -405,3 +420,27 @@ func matchersToString(ms ...*labels.Matcher) string {
...
+type QueryParseError struct {
+	msg string
+}
+
+func (e QueryParseError) Error() string {
+	return e.msg
+}
+
+func newQueryParseError(err error) QueryParseError {
+	return QueryParseError{msg: fmt.Sprintf("Error parsing query string %s", err.Error())}
```
Comment (squat) [inline injectproxy/routes.go:433]: Perhaps we should quote here to clarify where exactly the query starts and ends
```suggestion
	return QueryParseError{msg: fmt.Sprintf("Error parsing query string %q", err.Error())}
```

Comment (squat) [reply injectproxy/routes.go:433]: Error strings typically start with lower case letters so they can be wrapped:
```suggestion
	return QueryParseError{msg: fmt.Sprintf("error parsing query string %s", err.Error())}
```

File: injectproxy/routes.go [hunk 2]
```diff
@@ -405,3 +420,27 @@ func matchersToString(ms ...*labels.Matcher) string {
...
+type EnforceLabelError struct {
+	msg string
+}
+
+func (e EnforceLabelError) Error() string {
+	return e.msg
+}
+
+func newEnforceLabelError(err error) EnforceLabelError {
+	return EnforceLabelError{msg: fmt.Sprintf("Error enforcing label %s", err.Error())}
```
Comment (squat) [inline injectproxy/routes.go:445]: Here too:
```suggestion
	return EnforceLabelError{msg: fmt.Sprintf("error enforcing label %q", err.Error())}
```

File: injectproxy/routes.go [hunk 3]
```diff
@@ -405,3 +420,27 @@ func matchersToString(ms ...*labels.Matcher) string {
...
+
+func (e QueryParseError) Error() string {
+	return e.msg
+}
+
+func newQueryParseError(err error) QueryParseError {
+	return QueryParseError{msg: fmt.Sprintf("Error parsing query string %s", err.Error())}
+}
+
+type EnforceLabelError struct {
```
Comment (squat) [inline injectproxy/routes.go:436]: Does this need to be exported or can we keep it lower case? If it really must be exported, then please add comment to document it
```suggestion
type enforceLabelError struct {
```

File: injectproxy/routes.go [hunk 4]
```diff
@@ -405,3 +420,27 @@ func matchersToString(ms ...*labels.Matcher) string {
 	}
 	return fmt.Sprintf("{%v}", strings.Join(el, ","))
 }
+
+type QueryParseError struct {
```
Comment (squat) [inline injectproxy/routes.go:424]: Here too re: exporting
```suggestion
type queryParseError struct {
```

**Review by squat — COMMENTED** [review]
Thanks for doing this @prashbnair, it's long overdue. I just have a few comments

File: injectproxy/routes.go (hunk 1 above)
Comment (prashbnair) [reply injectproxy/routes.go]: Done

File: injectproxy/routes.go (hunk 2 above)
Comment (prashbnair) [reply injectproxy/routes.go]: Done

File: injectproxy/routes.go (hunk 3 above)
Comment (prashbnair) [reply injectproxy/routes.go]: Done

File: injectproxy/routes.go (hunk 4 above)
Comment (prashbnair) [reply injectproxy/routes.go]: Done

**Review by simonpasquier — APPROVED** [review]
lgtm

**Pull request merged**